### PR TITLE
[Kotlin Multiplatform] Update and fixes for Kotlin 1.5.10 and Kotlinx.serialization 1.2.1

### PR DESCRIPTION
**Pull request by 837**
- Update Kotlin Multiplatform with Kotlin 1.5.10, Ktor 1.6.0 and Kotlinx.serialization 1.2.1
- remove @serializable from interfaces
- Use CharArray.concatToString() instead of 'String(CharArray): String' as it is an error now
- Fix, RequestConfig needs a type to infer type variable T

Fixes from @DevSrSouzaern https://github.com/OpenAPITools/openapi-generator/pull/9232
Seemed not to be updated anymore.
- Fix `private` keyword not being handle as a reservedWords
- Fix enum serialization generation
- Migrate gradle build to Kotlin DSL

### PR checklist
 
- [X] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [X] Pull Request title clearly describes the work in the pull request and Pull Request description provides details about how to validate the work. Missing information here may result in delayed response from the community.
- [X] Run the following to [build the project](https://github.com/OpenAPITools/openapi-generator#14---build-projects) and update samples:
  ```
  ./mvnw clean package 
  ./bin/generate-samples.sh
  ./bin/utils/export_docs_generators.sh
  ``` 
  Commit all changed files. 
  This is important, as CI jobs will verify _all_ generator outputs of your HEAD commit as it would merge with master. 
  These must match the expectations made by your contribution. 
  You may regenerate an individual generator by passing the relevant config(s) as an argument to the script, for example `./bin/generate-samples.sh bin/configs/java*`. 
  For Windows users, please run the script in [Git BASH](https://gitforwindows.org/).
- [X] File the PR against the [correct branch](https://github.com/OpenAPITools/openapi-generator/wiki/Git-Branches): `master`, `5.1.x`, `6.0.x`
- [X] If your PR is targeting a particular programming language, @mention the [technical committee](https://github.com/openapitools/openapi-generator/#62---openapi-generator-technical-committee) members, so they are more likely to review the pull request.
@jimschubert , @dr4ke616, @karismann, @Zomzog, @andrewemery, @4brunu, @yutaka0m

### CODE REVIEW

**Comment by 837** [comment]
Someone an idea why CircleCi fails?

**Review by 4brunu — COMMENTED** [review]
Hi @837 since I don't have much experience with kotlin multiplatform I left some comments to try to understand the changes.

File: modules/openapi-generator/src/main/resources/kotlin-client/libraries/multiplatform/infrastructure/ApiClient.kt.mustache [hunk 1]
```diff
@@ -60,16 +51,9 @@ import {{packageName}}.auth.*
     {{/hasAuthMethods}}
 
     {{#nonPublicApi}}internal {{/nonPublicApi}}companion object {
-        protected val UNSAFE_HEADERS = listOf(HttpHeaders.ContentType)
+        const val BASE_URL = "{{{basePath}}}"
 
-        private fun setMappers(serializer: KotlinxSerializer) {
```
[the remark this answers is not in the document]
Comment (837) [reply modules/openapi-generator/src/main/resources/kotlin-client/libraries/multiplatform/infrastructure/ApiClient.kt.mustache:65]: This was used to Serialize/Deserialize JSON with List as root. Newer versions of kotlinx.serialization don't need this anymore. We can just pass `List<T>` and it returns the correct object.

File: modules/openapi-generator/src/main/resources/kotlin-client/libraries/multiplatform/api.mustache [hunk 2]
```diff
@@ -6,28 +6,20 @@ package {{apiPackage}}
...
-    serializer: KotlinxSerializer
-) : ApiClient(baseUrl, httpClientEngine, serializer) {
-
-    @UseExperimental(UnstableDefault::class)
-    constructor(
-        baseUrl: kotlin.String = "{{{basePath}}}",
-        httpClientEngine: HttpClientEngine? = null,
-        jsonConfiguration: JsonConfiguration = JsonConfiguration.Default
-    ) : this(baseUrl, httpClientEngine, KotlinxSerializer(Json(jsonConfiguration)))
+    jsonSerializer: Json
```
[the remark this answers is not in the document]
Comment (837) [reply modules/openapi-generator/src/main/resources/kotlin-client/libraries/multiplatform/api.mustache:21]: The creation of the KotlinxSerializer is now lazily done in the ApiClient::class.
The Json is used to pass down settings as they might vary depending on the api: 
```
Json {
            isLenient = true
            ignoreUnknownKeys = true
        }
```

File: modules/openapi-generator/src/main/resources/kotlin-client/libraries/multiplatform/api.mustache [hunk 3]
```diff
@@ -103,23 +95,5 @@ import kotlinx.serialization.internal.StringDescriptor
 {{/isMap}}
 
     {{/operation}}
-
-    {{#nonPublicApi}}internal {{/nonPublicApi}}companion object {
```
[the remark this answers is not in the document]
Comment (837) [reply modules/openapi-generator/src/main/resources/kotlin-client/libraries/multiplatform/api.mustache:107]: Same as a bit further down. https://github.com/OpenAPITools/openapi-generator/pull/9755/files#r651712318

We don't need to create mappers for Lists anymore.

**Comment by 4brunu** [comment]
I'm not a kotlin multiplaform users, but I don't see nothing wrong here.
@wing328 what do you think of this PR?

File: modules/openapi-generator/src/main/resources/kotlin-client/libraries/multiplatform/infrastructure/ApiClient.kt.mustache [hunk 4]
```diff
@@ -174,10 +158,10 @@ import {{packageName}}.auth.*
...
 
-        }.response
+        }
     }
 
-    private fun RequestConfig.updateForAuth(authNames: kotlin.collections.List<String>) {
+    private fun <T: Any?> RequestConfig<T>.updateForAuth(authNames: kotlin.collections.List<String>) {
         for (authName in authNames) {
             val auth = authentications?.get(authName) ?: throw Exception("Authentication undefined: $authName")
             auth.apply(query, headers)
```
Comment (krzema12) [inline modules/openapi-generator/src/main/resources/kotlin-client/libraries/multiplatform/infrastructure/ApiClient.kt.mustache:167]: A comment to line which mentions https://github.com/ktorio/ktor/issues/851 and the below workaround (I cannot place a comment there :disappointed:): how about removing this workaround? I see that this issue was resolved 2 years ago, so the newest version of Ktor almost for sure doesn't have this issue. On the other hand, to keep the number of changes minimal, I'm fine with doing it as a follow-up.

File: modules/openapi-generator/src/main/resources/kotlin-client/README.mustache [hunk 5]
```diff
@@ -3,11 +3,11 @@
 ## Requires
 
 {{#jvm}}
-* Kotlin 1.4.30
+* Kotlin 1.5.10
```
Comment (krzema12) [inline modules/openapi-generator/src/main/resources/kotlin-client/README.mustache:6]: I think the version for the `jvm` variant should remain unchanged. Otherwise it results in changing this version in READMEs for `kotlin-gson`, `kotlin-jvm-okhttp4-coroutines`, `kotlin-moshi-codegen` and other, and in this PR we only deal with the multiplatform variant.

File: modules/openapi-generator/src/main/resources/kotlin-client/libraries/multiplatform/api.mustache (hunk 2 above)
[the remark this answers is not in the document]
Comment (krzema12) [reply modules/openapi-generator/src/main/resources/kotlin-client/libraries/multiplatform/api.mustache:21]: Now when instantiating some `Api` class, it's needed to pass `jsonSerializer`. It wasn't needed before this PR (`jsonConfiguration: JsonConfiguration = JsonConfiguration.Default` in the constructor that was removed), so I think that we should remain backward-compatible here and some kind of a sane default `Json` configuration would be good. Best if this default configuration is accessible for the client, so that they can customize it by building upon the default one.

I agree that e.g. `ignoreUnknownKeys = true` should be in this default. Otherwise if the service being called starts including some new properties, JSON deserialization will fail. No experience/opinion on `isLenient`.

**Review by krzema12 — CHANGES_REQUESTED** [review]
Thanks for opening this PR! I started one myself (https://github.com/OpenAPITools/openapi-generator/pull/9827) but it's so similar to yours that I'm happy to help you polish this one :)

As a smoke test, I cloned your repo and branch, and generated a client for my company's internal service. For an example call, it works good!

Apart from the comments I placed inline, let me repeat concerns that I have for my PR:
* are there any end-to-end tests that actually try to call the Petstore service with the generated module? Ideally we'd like to test various edge cases in an automated and repeatable way
* how to ensure there are no regressions in other Kotlin templates that reuse the shared pieces of template?
* is `master` the right target branch for this change, given it's formally backward-incompatible? The policy is confusing to me

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/KotlinClientCodegen.java [hunk 6]
```diff
@@ -188,7 +188,7 @@ public KotlinClientCodegen() {
         supportedLibraries.put(JVM_OKHTTP4, "[DEFAULT] Platform: Java Virtual Machine. HTTP client: OkHttp 4.2.0 (Android 5.0+ and Java 8+). JSON processing: Moshi 1.8.0.");
         supportedLibraries.put(JVM_OKHTTP3, "Platform: Java Virtual Machine. HTTP client: OkHttp 3.12.4 (Android 2.3+ and Java 7+). JSON processing: Moshi 1.8.0.");
         supportedLibraries.put(JVM_RETROFIT2, "Platform: Java Virtual Machine. HTTP client: Retrofit 2.6.2.");
-        supportedLibraries.put(MULTIPLATFORM, "Platform: Kotlin multiplatform. HTTP client: Ktor 1.2.4. JSON processing: Kotlinx Serialization: 0.12.0.");
+        supportedLibraries.put(MULTIPLATFORM, "Platform: Kotlin multiplatform. HTTP client: Ktor 1.5.3. JSON processing: Kotlinx Serialization: 1.1.0.");
```
[the remark this answers is not in the document]
Comment (837) [reply modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/KotlinClientCodegen.java]: True, forgot to change them after upgrading the on ein the build.gradle

File: modules/openapi-generator/src/main/resources/kotlin-client/data_class.mustache [hunk 7]
```diff
@@ -69,9 +70,9 @@ import java.io.Serializable
      * {{{description}}}
      * Values: {{#allowableValues}}{{#enumVars}}{{&name}}{{^-last}},{{/-last}}{{/enumVars}}{{/allowableValues}}
      */
-    {{#kotlinx_serialization}}
```
[the remark this answers is not in the document]
Comment (837) [reply modules/openapi-generator/src/main/resources/kotlin-client/data_class.mustache:72]: Wasn't that only used in the context of multiplatform?

[the remark this answers is not in the document]
Comment (krzema12) [reply modules/openapi-generator/src/main/resources/kotlin-client/data_class.mustache:72]: I'm not sure actually, I assumed that `kotlinx_serialization` is a separate library. But if it was just a flag in multiplatform, then looks good :+1:

[the remark this answers is not in the document]
Comment (4brunu) [reply modules/openapi-generator/src/main/resources/kotlin-client/data_class.mustache:72]: `kotlinx_serialization` is also used in the kotlin client jvm.
Could you please revert this change?

[the remark this answers is not in the document]
Comment (837) [reply modules/openapi-generator/src/main/resources/kotlin-client/data_class.mustache:72]: Will do!

File: modules/openapi-generator/src/main/resources/kotlin-client/libraries/multiplatform/infrastructure/ApiClient.kt.mustache [hunk 8]
```diff
@@ -60,16 +51,9 @@ import {{packageName}}.auth.*
     {{/hasAuthMethods}}
 
     {{#nonPublicApi}}internal {{/nonPublicApi}}companion object {
+        const val BASE_URL = "{{{basePath}}}"
+        const val JSON_DEFAULT = Json { ignoreUnknownKeys = true }
```
Comment (krzema12) [inline modules/openapi-generator/src/main/resources/kotlin-client/libraries/multiplatform/infrastructure/ApiClient.kt.mustache:55]:
```
Const 'val' has type 'Json'. Only primitives and String are allowed 
```
Some tests that would make CI catch it would be really good to have :sweat_smile:

File: samples/client/petstore/kotlin-multiplatform/src/commonMain/kotlin/org/openapitools/client/infrastructure/ApiClient.kt [hunk 9]
```diff
@@ -53,23 +44,9 @@ open class ApiClient(
     }
 
     companion object {
+        const val BASE_URL = "http://petstore.swagger.io/v2"
+        const val JSON_DEFAULT = Json { ignoreUnknownKeys = true }
```
Comment (krzema12) [inline samples/client/petstore/kotlin-multiplatform/src/commonMain/kotlin/org/openapitools/client/infrastructure/ApiClient.kt:48]: This forbidden `const` got also in the sample. I thought CI tries to at least build the sample projects?

**Review by krzema12 — CHANGES_REQUESTED** [review]

File: samples/client/petstore/kotlin-multiplatform/src/commonMain/kotlin/org/openapitools/client/infrastructure/ApiClient.kt (hunk 9 above)
Comment (4brunu) [reply samples/client/petstore/kotlin-multiplatform/src/commonMain/kotlin/org/openapitools/client/infrastructure/ApiClient.kt:48]: Maybe not the multiplaform sample projects, but the jvm kotlin client samples are build in CI.

File: modules/openapi-generator/src/main/resources/kotlin-client/libraries/multiplatform/infrastructure/ApiClient.kt.mustache (hunk 8 above)
Comment (837) [reply modules/openapi-generator/src/main/resources/kotlin-client/libraries/multiplatform/infrastructure/ApiClient.kt.mustache:55]: Oh man, obviously 🙄 Well spotted

**Review by krzema12 — APPROVED** [review]
LGTM - I think we can merge as is because it works fine, and then apply any improvements like end-to-end tests or code cleanup as follow-ups.

**Comment by 4brunu** [comment]
@837 could you please close and reopen this PR to restart CI? Thanks

**Comment by 837** [comment]
sure

**Comment by 837** [comment]
https://github.com/OpenAPITools/openapi-generator/pull/9862
Reopened as per request of @4brunu

**Comment by 4brunu** [comment]
Sorry if I miscommunicated, but to restart the CI you could just close and reopen the same PR, there is no need to create a new one.
But this works too 🙂

**Comment by 837** [comment]
Oh didnt know that. Makes more sense to keep this one.

**Comment by 4brunu** [comment]
Looks good to me 👍

**Comment by wing328** [comment]
@837 can you please resolve the merge conflicts when you've time?

**Comment by wing328** [comment]
Merged into master and updated samples via 173c7ca652d

**Comment by wing328** [comment]
Kotlin tests pass via https://github.com/OpenAPITools/openapi-generator/pull/9883

**Comment by shadowsheep1** [comment]
I had to use a custom mustache template for `HttpBasicAuth.kt.mustache` to change it as:

```mustache
package {{packageName}}.auth

import io.ktor.util.InternalAPI
import io.ktor.util.encodeBase64

/**
 * Custom Template
 */

class HttpBasicAuth : Authentication {
    var username: String? = null
    var password: String? = null

    @OptIn(InternalAPI::class)
    override fun apply(query: MutableMap<String, List<String>>, headers: MutableMap<String, String>) {
        if (username == null && password == null) return
        val str = (username ?: "") + ":" + (password ?: "")
        val auth = str.encodeBase64()
        headers["Authorization"] = "Basic $auth"
    }
}

```
Because with AS Artic Fox 2020.3.1 Patch 2

With this configuration

```kotlin
object Versions {
    const val iosTestTask = false
    const val foo = "8.6.8"
    const val kotlin = "1.5.31"

    // Android
    const val compileSdk = 30
    const val minSdk = 23
    const val targetSdk = compileSdk

    // You can also use this to specify versions for dependencies. Having consistent
    // versions between modules can avoid behavior conflicts.
    const val sqldelight = "1.5.1"
    const val settings = "0.8"
    const val serialization = "1.2.2"
    // https://kotlinlang.org/docs/mobile/concurrency-and-coroutines.html#multithreaded-coroutines
    const val coroutines = "1.5.1-native-mt"
    const val ktor = "1.6.3"
    const val openapigen = "5.2.1"
}
```
I had (using `@ExperimentaAPI`) annotation this error.

```shell
e: Opt-in requirement marker annotation on override requires the same marker on base declaration
```

I'd to use `@OptIn(InternalAPI::class)` instead

**Comment by 4brunu** [comment]
@shadowsheep1 could you please add a PR with those changes?

**Comment by shadowsheep1** [comment]
@4brunu sure 👍

**Comment by shadowsheep1** [comment]
@4brunu do you have a particular PR format to follow?
